src-input/duk_bi_json.c

```diff
@@ -1638,12 +1641,13 @@ DUK_LOCAL void duk__enc_pointer(duk_json_enc_ctx *js_ctx, void *ptr) {
 #if defined(DUK_USE_JC)
 	{
 		DUK_ASSERT(js_ctx->flag_ext_compatible);
-		fmt = ptr ? "{\"_ptr\":\"%p\"}" : "{\"_ptr\":\"null\"}";
+		fmt = ptr ? "{\"_ptr\":\"%s\"}" : "{\"_ptr\":\"null\"}";
 	}
 #endif
+	duk_encode_pointer_cstr(ptrbuf, sizeof(ptrbuf), ptr);
 
 	/* When ptr == NULL, the format argument is unused. */
-	DUK_SNPRINTF(buf, sizeof(buf) - 1, fmt, ptr);  /* must not truncate */
+	DUK_SNPRINTF(buf, sizeof(buf) - 1, fmt, ptrbuf);  /* must not truncate */
```

Comment:
Sidenote: Duktape internals don't currently assume the C99 behavior for truncation (= guaranteed NUL unless bufsize is 0) because some snprintf() variants (like MSVC `_snprintf()`) do not guarantee NUL termination. So call sites either apply NUL termination explicitly, or use (size - 1) on a zeroed buffer to guarantee the termination.

IMO the best way forward would be to require DUK_SNPRINTF() to have C99 truncation behavior. This would remove the need for using (size - 1) or explicit NUL termination.

I think I'll do a pass on the snprintf() call sites after this pull.